phoenix-core/src/main/java/org/apache/phoenix/expression/LiteralExpression.java

```diff
@@ -54,184 +54,218 @@
-                type.toBytes(value, sortOrder);
-        if (type == PVarchar.INSTANCE || type == PChar.INSTANCE) {
-            if (type == PChar.INSTANCE && maxLength != null  && b.length < maxLength) {
-                if (rowKeyOrderOptimizable) {
-                    b = type.pad(b, maxLength, sortOrder);
-                } else {
-                    b = StringUtil.padChar(b, maxLength);
+
+        public LiteralExpression buildValueAndDeterminism(boolean byteValueOnly) {
+            /***
+             * @param byteValueOnly: true if building expression that only has byteValue, false otherwise
+             */
+            //for when you only have value (could be null) and determinism (can also be null)
+            // all other fields should be null
+            if (byteValueOnly) {
+                return new LiteralExpression(null, null, this.byteValue = this.byteValue != null ? this.byteValue : ByteUtil.EMPTY_BYTE_ARRAY, null, null, SortOrder.ASC, Determinism.ALWAYS);
+            }
+            setDefaults();
+            if (this.value instanceof Boolean) {
+            return LiteralExpression.getBooleanLiteralExpression((Boolean)this.value, this.determinism);
+            }
+            else if (this.value == null) {
+                return LiteralExpression.getNullLiteralExpression(this.determinism);
+            }
+            PDataType type = PDataType.fromLiteral(this.value);
+            this.byteValue = type.toBytes(this.value);
+            if (type.isNull(this.byteValue)) {
+                return LiteralExpression.getTypedNullLiteralExpression(type, this.determinism);
+            }
+            if (type == PVarchar.INSTANCE) {
+                String s = (String) this.value;
+                if (s.length() == this.byteValue.length) { // single byte characters only
+                    type = PChar.INSTANCE;
                 }
-            } else if (value != null) {
-                maxLength = ((String)value).length();
             }
-        } else if (type.isArrayType()) {
-            maxLength = ((PhoenixArray)value).getMaxLength();
-        }
-        if (b.length == 0) {
-            return getTypedNullLiteralExpression(type, determinism);
+            return new LiteralExpression(this.value, type, this.byteValue, type == null || !type.isFixedWidth() ? null :
+                    type.getMaxLength(this.value), null, SortOrder.getDefault(), this.determinism);
         }
-        if (maxLength == null) {
-            maxLength = type.isFixedWidth() ? type.getMaxLength(value) : null;
+
+        public LiteralExpression build() throws SQLException{
+            setDefaults();
+            if (this.value == null) {
+                return  (this.type == null) ?  LiteralExpression.getNullLiteralExpression(this.determinism) :
+                        LiteralExpression.getTypedNullLiteralExpression(this.type, this.determinism);
+            }
+            else if (this.value instanceof Boolean){
```

Comment:
nit: inconsistent coding style, most of the time you keep `else if` with the same line after `}`, maybe follow the same coding style here?